### ISSUE TITLE
Packages from `nodePackages_latest` don't result in executables in $PATH.

### ISSUE THREAD

**Issue by seanparsons** [NixOS/nixpkgs#145432]
### Describe the bug
A shell including packages from `nodePackages_latest` like `yarn` or `pnpm` don't result in executables on the path the way `nodePackages` does.

### Steps To Reproduce
1. Run `nix-shell -p nodePackages_latest.yarn --run "yarn --version"`

### Expected behavior
I would expect to see the version number of yarn instead of `/tmp/nix-shell-433770-0/rc: line 1: yarn: command not found`.

### Additional context
This works find if `nix-shell -p nodePackages.yarn --run "yarn --version"` is run instead.

I stumbled on this earlier but I'm not entirely certain if it's related: https://github.com/NixOS/nixpkgs/pull/142915

### Notify maintainers
Apologies @marsam if you're not the most appropriate person to pull into this!

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.76, NixOS, 21.05.3990.372e59d2af7 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.05.3990.372e59d2af7, nixos-hardware"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by seanparsons** [comment]
I've also attempted this with a more bleeding edge version of nixpkgs and it still appears to be an issue.

**Comment by andersk** [comment]
This seems to be a node2nix issue. Cc @svanderburg.

**Comment by happysalada** [comment]
node2nix only supports node_16 on master, the version included in nixpkgs does not support it.
Perhaps we could do a bump to unstable for node2nix ?

**Comment by andersk** [comment]
node2nix `master` still has this issue.

**Comment by miuirussia** [comment]
> node2nix `master` still has this issue.

confirm it

**Comment by CMCDragonkai** [comment]
Yes node2nix master has this issue: https://github.com/svanderburg/node2nix/issues/293.

However I'm not sure if this has anything to do with the OP's problem.

This is because nixpkgs master isn't even using 1.10.0 or 1.11.0: https://github.com/svanderburg/node2nix/issues/291#issuecomment-1108130555

**Comment by andersk** [comment]
@CMCDragonkai I suspect your test on 1.9.0 was tainted by the fact that 1.9.0 did not support `--nodejs-16` (svanderburg/node2nix@e73c616408b753c386301e1d21ac634c959a4b6f), and you didn’t notice because 1.9.0 also did not complain about invalid options (svanderburg/node2nix#272). If you account for this ([`--pkg-name nodejs-16_x`](https://github.com/svanderburg/node2nix/issues/236#issuecomment-947465067)), you’ll see that 1.9.0 is equally affected, matching the OP’s problem.

**Comment by CMCDragonkai** [comment]
I'm not sure what OP's problem actually is because I'm not sure what the current nixpkg's master node packages are built with. As I said, the master claims that it has 1.10.0, but that package **is actually** 1.9.0. And there was a PR to update all of the packages but that isn't even merged.

My experiments show that a combination of behaviour changes with `npm install` and `node2nix` has resulted it in no longer making the bin executable symlinks. So I'm not sure what the fix will be.

While we wait on upstream to figure it out, I've ended up with duct-tape solution using `jq` to directly parse out the package.json information to produce executable symlinks: https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/37#issuecomment-1108639198

**Comment by andersk** [comment]
@CMCDragonkai I’m sure. `master` has 1.9.0 with some patches, see this override:

https://github.com/NixOS/nixpkgs/blob/87d34a6b8982e901b8e50096b8e79ebc0e66cda0/pkgs/development/node-packages/default.nix#L251-L294

It is easy to verify that 1.9.0 is equally affected without relying on nixpkgs:

```console
$ echo `["yarn"]` > node-packages.json

$ npx node2nix@1.9.0 --version
Need to install the following packages:
  node2nix@1.9.0
Ok to proceed? (y) y
node2nix 1.9.0

$ npx node2nix@1.9.0 --pkg-name nodejs-16_x --input node-packages.json 
info attempt registry request try #1 at 12:10:57 PM
http request GET https://registry.npmjs.org/yarn
http 200 https://registry.npmjs.org/yarn

$ nix-build -A yarn
…

$ ls result/
lib

$ npx node2nix@1.9.0 --pkg-name nodejs-14_x --input node-packages.json 
info attempt registry request try #1 at 12:14:27 PM
http request GET https://registry.npmjs.org/yarn
http 200 https://registry.npmjs.org/yarn

$ nix-build -A yarn
…

$ ls result/
bin  lib

```

Same results with 1.9.0, 1.10.0, and 1.11.0. It really is a difference between `--pkg-name nodejs-14_x` and `--pkg-name nodejs-16_x`, not a difference between node2nix versions.

In nixpkgs, it’s the difference between `nodePackages` and `nodePackages_latest` for the same reason, because `nodePackages` is defined with `.override { nodejs = nodejs-14_x; }`, while `nodePackages_latest` gets `nodejs_latest` which is `nodejs-16_x`.

**Comment by lilyinstarlight** [comment]
This should be fixed by #193337, if anyone would like to test (`nodePackages.yarn` at least seems to work with `nodejs-18_x` on my system now)

**Comment by utybo** [comment]
I can confirm this is fixed, I no longer need to create symlinks manually for node packages, thank you very much to the devs who fixed this!
